### ISSUE TITLE
`Qwen3_5GatedDeltaNet` / `Qwen3_5MoeGatedDeltaNet` initialize `A_log` to `-inf` on some heads (bf16 models)

### ISSUE THREAD

**Issue by Nkluge-correa** [huggingface/transformers#47831]
### System Info

- `transformers` version: 5.14.0
- Platform: Linux (glibc 2.34)
- Python version: 3.12.5
- PyTorch version (GPU?): 2.13.0+cu126 (yes, 8× NVIDIA H200)
- Model: Qwen3.5-MoE hybrid (40 layers = 30 GatedDeltaNet + 10 full attention, 256 experts), trained from scratch in bf16

Output of `transformers env` is consistent with the above.

### Who can help?

Text/hybrid (linear-attention) models — Qwen3.5 GatedDeltaNet (@ArthurZucker @Cyrilvallez @vasqu)

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

When a Qwen3.5 (`qwen3_5`) or Qwen3.5-MoE (`qwen3_5_moe`) model is instantiated in
bfloat16 (the normal case for `from_config(..., dtype=torch.bfloat16)` or
`from_pretrained(..., dtype=torch.bfloat16)` on a fresh config), a fraction of the
GatedDeltaNet heads get `A_log = -inf`, i.e. `A = exp(A_log) = 0`.

The cause is the decay-parameter initialization in
`Qwen3_5GatedDeltaNet.__init__` / `Qwen3_5MoeGatedDeltaNet.__init__`:

```python
A = torch.empty(self.num_v_heads).uniform_(0, 16)   # inherits model dtype (bf16)
self.A_log = nn.Parameter(torch.log(A))
```

`torch.empty(self.num_v_heads)` inherits the module's default dtype. Under a bf16
model that is **bfloat16**, where `uniform_(0, 16)` rounds a draw down to exactly
`0.0` with probability ≈ 0.4% per head. `torch.log(0) = -inf`, so that head's
decay `A = exp(A_log) = 0`.

Consequences for the affected head:
- `g = -A * softplus(a + dt_bias) = 0` -> the gated-delta recurrence **never
  decays/forgets** for that head (its state accumulates without bound).
- `dg/dA_log = -A * softplus(...) = 0` -> the gradient w.r.t. `A_log` is exactly 0,
  so the head is **permanently frozen at `-inf`** and can never recover.

Because the probability compounds over `num_value_heads × num_gdn_layers`, small
models are usually spared while large hybrids are reliably affected. On our 35B-A3B
config (32 value heads × 30 GDN layers ≈ 960 draws) we observe ~3–6 dead heads at
every initialization; an 8×8 tiny config shows none.

## Reproduction

```python
import torch

torch.manual_seed(0)
num_value_heads, num_gdn_layers = 32, 30

def count_dead(dtype):
    dead = 0
    for _ in range(num_gdn_layers):
        A = torch.empty(num_value_heads, dtype=dtype).uniform_(0, 16)  # current code
        A_log = torch.log(A.float()) if dtype == torch.float16 else torch.log(A)
        dead += int((~torch.isfinite(A_log)).sum())
    return dead

print("bfloat16 dead heads:", count_dead(torch.bfloat16))  # e.g. 6
print("float32  dead heads:", count_dead(torch.float32))   # 0
```

Or directly on a model:

```python
import torch
from transformers import AutoModelForCausalLM, AutoConfig

config = AutoConfig.from_pretrained("<a qwen3_5_moe hybrid config>")
model = AutoModelForCausalLM.from_config(config, dtype=torch.bfloat16)

for name, p in model.named_parameters():
    if name.endswith("A_log") and not torch.isfinite(p).all():
        print(name, "->", int((~torch.isfinite(p)).sum()), "dead head(s)")
```

### Expected behavior

`A_log` should be finite for every head at initialization, regardless of the model
dtype. No head should start with a zero (never-forgetting) decay gate.

## Suggested fix

Sample `A` in float32 before taking the log, matching the reference
implementations already in the repo. This is a 1-line change in each of
`Qwen3_5GatedDeltaNet.__init__` and `Qwen3_5MoeGatedDeltaNet.__init__`:

```diff
- A = torch.empty(self.num_v_heads).uniform_(0, 16)
+ A = torch.empty(self.num_v_heads, dtype=torch.float32).uniform_(0, 16)
  self.A_log = nn.Parameter(torch.log(A))
```

This aligns with existing models in the codebase, which already build `A` in
float32:
- `Mamba2Mixer` (`modeling_mamba2.py`)
- `Qwen3NextGatedDeltaNet` (`modeling_qwen3_next.py`)

so the fix keeps Qwen3.5 consistent with the rest of the library rather than
diverging. The `dt_bias`/`A_log` parameters are also intended to be decay-free;
keeping them float32 additionally avoids losing precision on the decay term.

I'm happy to open the PR with this change if it looks good to the maintainers.

**Comment by navaneethkrishnansuresh** [comment]
Hi, I'd like to take this if it's free.

Rough fix: sample A in float32 before log (so bf16 can't round to 0 → -inf), apply the same in the GatedDeltaNet _init_weights path, keep modular/generated files in sync, and add a regression test that A_log stays finite under bf16.

Will open a PR once tests pass.

**Comment by navaneethkrishnansuresh** [comment]
I've opened a PR with the float32 `A` sample fix for GDN init / `_init_weights` (Qwen3Next + Qwen3.5 / MoE) and a regression test:

https://github.com/huggingface/transformers/pull/47842

**Comment by Nkluge-correa** [comment]
> *"I'm happy to open the PR with this change if it looks good to the maintainers."*

But thanks for jumping in front of something I have been investigating for a week @navaneethkrishnansuresh 🫠

**Comment by navaneethkrishnansuresh** [comment]
Oops, didn't see this @Nkluge-correa  😅

**Comment by Rocketknight1** [comment]
hey, this bug looks real, although it only affects fresh initializations and not pretrained checkpoints. still, i'm happy to wait for the PR from @Nkluge-correa! Everyone else please stop running your code agents on this or I'll block you

**Comment by Nkluge-correa** [comment]
Thanks for the opportunity to contribute 🤗
Just made my PR. Please let me know if anything doesn't look good.
